File: Data Structures & Algorithms/rotate-matrix/submission-0.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length - 1;
        int l = 0;
        int r = n;
        int t = 0;
        int b = n;

        while (l < r) {
            int i = 0;
            while (i < r - l) {
                int temp = matrix[l+i][t];
                matrix[l+i][t] = matrix[r][t+i];
                matrix[r][t+i] = matrix[r-i][b];
                matrix[r-i][b] = matrix[l][b-i];
                matrix[l][b-i] = temp;
                i++;
            }
            l++;
            r--;
            t++;
            b--;
        }
    }
    
}

File: Data Structures & Algorithms/search-2d-matrix/submission-0.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        w = len(matrix[0])
        l = 0
        r = w * len(matrix) - 1
        while (l < r):
            mid = l + (r-l)//2
            print(f"i{mid//w}, j{mid%w}")
            if target < matrix[mid//w][mid%w]:
                r = mid - 1
            elif target > matrix[mid//w][mid%w]:
                l = mid + 1
            else:
                return True
        return (matrix[l//w][l%w] == target)
